src/module/system/damage/weapon.ts

```diff
@@ -315,14 +315,14 @@ class WeaponDamagePF2e {
             // If an alternate critical specialization effect is available, apply it only if there is also a
             // qualifying non-alternate
             const critSpecs = actor.synthetics.criticalSpecalizations;
-            const standard = critSpecs.standard
-                .map((cs) => cs(weapon, options))
-                .filter((ce): ce is CritSpecEffect => !!ce)
-                .pop();
-            const alternate = critSpecs.alternate
-                .map((cs) => cs(weapon, options))
-                .filter((ce): ce is CritSpecEffect => !!ce)
-                .pop();
+            const standard = critSpecs.standard.reduceRight(
+                (res, cs) => res ?? cs?.(weapon, options),
+                null as CritSpecEffect | null
+            );
+            const alternate = critSpecs.alternate.reduceRight(
```

Comment:
You mean `res`?  It was short for result.  Would `result` be ok?

While it can't short circuit the reduce, it does stop evaluating the cs functions after finding the first (last) one that works, so all the predicate evaluation is avoided.  I thought about using findLast, but then the found cs must be evaluated twice, which will be worse.